### ISSUE TITLE
Gentoo hangs during toolchain-config.in for arm-unknown-eabi

### ISSUE THREAD

**Issue by ntzrmtthihu777** [crosstool-ng/crosstool-ng#538]
As the title states. Since gentoo's crossdev tool is not cooperating with me in building the equivalent of 'arm-none-eabi' as packaged by arch linux and other distros I decided to give ct-ng a try.
Aside from the fact that apparently `LD_LIBRARY_PATH` is set to `/usr/lib` by default on gentoo (I don't recall ever setting this explicitly, so I assume its a default for now), unsetting the variable and running
`ct-ng arm-unknown-eabi && ct-ng build` hangs for an apparently infinite amount of time at the following step according to build.log:

`[DEBUG]  ==> Executing: '/usr/lib/portage/python3.4/ebuild-helpers/xattr/install' '-c' '-m' '0755' '/usr/lib64/crosstool-ng-1.22.0/scripts/toolchain-config.in' '/home/netz/x-tools/arm-unknown-eabi/bin/arm-unknown-eabi-ct-ng.config'`

Unsure how to proceed or debug. Most I found googling this string was some chatter on a russian gento irc channel.

**Comment by jasmin-j** [comment]
Hi,
I was working quite some time on this issue, but due to lack if time it isn't finished now.
You may look to pull request #372 and use it as a start point.
The goal of #372 is to get **exactly** the same compiler, as the ARM embedded branch does create, but using CT-NG instead of the ARM scripts.

CT-NG evolved really a lot since my work and may be able to build it now. I guess to finish it, there is also PR #375 to be done first. AFAIK the compiler built already, but GDB was a problem to be built on MinGW. The latter really improved a lot since my tries.

I am on vacation for longer time and have only limited iNet access, so you may finish my work sooner than I can.

BR,
Jasmin

**Comment by stilor** [comment]
I am not a gentoo user, but quick googling shows this to be some gentoo issue with this replacement for install. E.g. https://bugs.funtoo.org/browse/FL-2780 reports 'install' was hanging for 42 minutes doing *something*.

Can you run the same command manually? Is anything in the system log - e.g. some security policy being checked by Gentoo, etc?

I am not sure what the gentoo is doing in that install and why it takes so long. But it looks like gentoo's bug, not crosstool-ng's.

**Comment by ntzrmtthihu777** [comment]
Good find, I'll have a look into it once I get back to my machine.

**Comment by ntzrmtthihu777** [comment]
@stilor running the above command manually does in fact work. No logs saying 'nuh-uh' I can find.

**Comment by stilor** [comment]
Well:
1. The configure detects `/usr/bin/install` for me, not `/usr/lib/portage/python3.4/ebuild-helpers/xattr/install`. How was your crosstool-ng configured and installed?
2. If I configure crosstool-ng passing `--with-install=/usr/lib/portage/python3.4/ebuild-helpers/xattr/install`, I see the same bug.

I looked into that script and I now know what's going on: the script invokes `/usr/bin/install-xattr` wrapper, passing its own path in a special environment variable (`__PORTAGE_HELPER_PATH`). `install-xattr` in turn tries to find real `install` program by scanning all directories in `$PATH` and excluding itself (as obtained by reading `/proc/self/exe`) and portage helper, if any, as obtained from that environment variable.

The problem is that crosstool-ng adds an internal directory into `$PATH`, `.build/tools/bin` where it create its own wrapper for `install`. Unsurprisingly, this points back to `/usr/lib/portage/python3.4/ebuild-helpers/xattr/install`.

In other words, the portage helper and install-xattr are capable of avoiding a loop in a two-step invocation, but crosstool-ng creates a third entry in the loop: `/usr/lib/portage/python3.4/ebuild-helpers/xattr/install` -> `/usr/bin/install-xattr` -> `.build/tools/bin/install` -> `/usr/lib/portage/python3.4/ebuild-helpers/xattr/install`.

As a workaround, configure crosstool-ng with --with-install=/usr/bin/install and you should be fine. I'll think how to work around this braindamaged loop in install wrappers...

@PS @jasmin-j Let's keep this issue for the build issue on gentoo.

**Comment by stilor** [comment]
Should be fixed on master.
